rfcs/0029-smart-contracts.md

```diff
@@ -0,0 +1,420 @@
+      significantly harder to track gas usage, and more expensive to reset to a
+      known-good environment.
+
+## Drawbacks and benefits
+[drawbacks-and-benefits]: #drawbacks-and-benefits
+
+### Witness generation + proving off chain: store verification keys in accounts
+
+#### Drawbacks
+
+* Proofs need to be generated by the sender of the transaction.
+  - This can be expensive, and may restrict the type of devices that can
+    send transactions that are verified by smart contracts.
+* Keys need to be generated by the contract author.
+  - This is also expensive, which will restrict the type of devices that can
+    set up a smart contract.
+* The smart contract is opaque.
+  - The verification key doesn't provide any insight into the constraint system
+    that the transaction must satisfy.
+  - The contents of the smart contract must be communicated to authorised
+    senders off-chain.
+    + This also involves transmitting the proving key, which can be large.
+* Keys should be regenerated per-account, reuse should be avoided.
+  - Re-using keys would allow the original key generator to generate false
+    proofs, if they had stored the 'toxic waste' randomness used when
+    generating them.
+
+#### Benefits
+
+* Least amount of work to implement.
+  - Most of the complex parts (the snarky verifier, account look-ups, etc.) are
+    already implemented and used elsewhere.
+* Verification keys will be fairly small, so storage costs can be lower.
+* Developers are not restricted to any particular language for generating their
+  keys and proofs.
+* The size of the constraint system is unrestricted.
+
+### Witness generation + proving off chain: store WASM blob for generating the constraints
+
+#### Drawbacks
+[constraints-only-wasm-drawbacks]: #drawbacks-1
+
+* Proofs need to be generated by the sender of the transaction.
+  - This can be expensive, and may restrict the type of devices that can
+    send transactions that are verified by smart contracts.
+* The contract author needs to compute the constraint system to generate the
+  appropriate hash.
+  - This should be fairly cheap, but may still require a local WASM
+    interpreter if the hash is not already known.
+* Need to implement a universal snark.
+* Need to implement a gas model.
+  - It is also not possible to verify the gas usage in the snark, so more trust
+    is placed in consensus.
+* Need to implement or integrate a WASM interpreter.
+  - Gas metering may require wide-reaching changes if we adapt an existing
+    interpreter.
+* Consensus nodes need to do an amount of unpaid work to verify the gas usage.
+* Constraint system programs must be written in a language that compiles down
+  to WASM.
+  - This usually means that the language relies on manual memory management,
+    which will make it hard to use for some developers.
+* Will need to write WASM backends for our current snark langauge tooling, if
```

Comment:
Do you have an estimate as to what this + writing the language would look like time wise? Maybe there is something off the shelf we could already use for a first version?